Feature/vudrm 1747/bento4 docs (#94)
Vualto/vudrm-public-docs: docs/source/DeveloperDocumentation/PackagingGuides/bento4.md

# Bento4

## Installing Bento4

You can download the binaries for Bento4 for your preferred OS at the following link - [Bento4 binaries](https://www.bento4.com/downloads/)

Add the filepath for the binaries to the Bento4 `bin` folder as a value to your `Path` environemnt variables on your computer.

## Get CPIX Documents

CPIX Documents are required to give Bento4 the necessary information in order to encrypt your media.

You can retrieve these documents from the VUDRM Admin site. 

A guide can be found at the following link - [VUDRM Admin - CPIX Document Request](https://docs.vualto.com/projects/vudrm/en/latest/UserGuide/VUDRM-Admin.html#vudrm-encryption-keys)

### Examples of Each Document

#### CPIX Document

Each DRM Provider has a unique `systemId`. This helps to distinguish which values belong to which DRM Provider in a CPIX Document. 

```

> System IDs:

Author: Actioned.

            <cpix:Data>
                <pskc:Secret>
                    <pskc:PlainValue>3wFFjPspsydY/t2uSPeE6w==</pskc:PlainValue>
                </pskc:Secret>
            </cpix:Data>
        </cpix:ContentKey>
    </cpix:ContentKeyList>
    <cpix:DRMSystemList>
        <cpix:DRMSystem kid="11111111-1111-1111-1111-111111111111" systemId="9a04f079-9840-4286-ab92-e65be0885f95">
            <cpix:PSSH><PlayReady_PSSH></cpix:PSSH>
        </cpix:DRMSystem>
        <cpix:DRMSystem kid="22222222-2222-2222-22222-22222222222" systemId="edef8ba9-79d6-4ace-a3c8-27dcd51d21ed">
            <cpix:PSSH><Widevine_PSSH></cpix:PSSH>
        </cpix:DRMSystem>
        <cpix:DRMSystem kid="33333333-3333-3333-33333-33333333333" systemId="94ce86fb-07ff-4f43-adb8-93d2fa968ca2">
            <cpix:URIExtXKey><FairPlay_URIExtXKey</cpix:URIExtXKey>
            <cpix:HLSSignalingData playlist="master"><master_playlist_value></cpix:HLSSignalingData>
            <cpix:HLSSignalingData playlist="media"><media_playlist_value></cpix:HLSSignalingData>
        </cpix:DRMSystem>
    </cpix:DRMSystemList>
    <cpix:ContentKeyUsageRuleList></cpix:ContentKeyUsageRuleList>
</cpix:CPIX>
```

#### CPIX keys as JSON

> Below is an example of a CPIX keys as JSON document with all `DRM Providers` selected. 

Author: Actioned.

# Bento4

## Installing Bento4

> You can download the binaries for Bento4 for your preferred OS at the following link - [Bento4 binaries](https://www.bento4.com/downloads/)

Reviewer: General website standards, from places like [w3](https://www.w3.org/QA/Tips/noClickHere#:~:text=Don't%20use%20%22click%20here%22%20as%20link%20text&text=When%20calling%20the%20user%20to,doesn't%20talk%20about%20mechanics), are against linking like this.
Something like the follow would be nicer:
```suggestion
You can download the binaries for your preferred OS from [Bento4](https://www.bento4.com/downloads/).
```

Add the filepath for the binaries to the Bento4 `bin` folder as a value to your `Path` environemnt variables on your computer.

## Get CPIX Documents

CPIX Documents are required to give Bento4 the necessary information in order to encrypt your media.

You can retrieve these documents from the VUDRM Admin site. 

> A guide can be found at the following link - [VUDRM Admin - CPIX Document Request](https://docs.vualto.com/projects/vudrm/en/latest/UserGuide/VUDRM-Admin.html#vudrm-encryption-keys)

Reviewer: As above.

Author: Actioned.

> More information on our CPIX Key Provider API can be found at the following link - [CPIX Key Provider API](https://docs.vualto.com/projects/vudrm/en/latest/DeveloperDocumentation/VUDRM-key-provision.html#cpix-key-provider-api)

Reviewer: And again.

Author: Actioned.

    </cpix:ContentKeyList>
    <cpix:DRMSystemList>
        <cpix:DRMSystem kid="11111111-1111-1111-1111-111111111111" systemId="9a04f079-9840-4286-ab92-e65be0885f95">
            <cpix:PSSH><PlayReady_PSSH></cpix:PSSH>
        </cpix:DRMSystem>
        <cpix:DRMSystem kid="22222222-2222-2222-22222-22222222222" systemId="edef8ba9-79d6-4ace-a3c8-27dcd51d21ed">
            <cpix:PSSH><Widevine_PSSH></cpix:PSSH>
        </cpix:DRMSystem>
        <cpix:DRMSystem kid="33333333-3333-3333-33333-33333333333" systemId="94ce86fb-07ff-4f43-adb8-93d2fa968ca2">
            <cpix:URIExtXKey><FairPlay_URIExtXKey</cpix:URIExtXKey>
            <cpix:HLSSignalingData playlist="master"><master_playlist_value></cpix:HLSSignalingData>
            <cpix:HLSSignalingData playlist="media"><media_playlist_value></cpix:HLSSignalingData>
        </cpix:DRMSystem>
    </cpix:DRMSystemList>
    <cpix:ContentKeyUsageRuleList></cpix:ContentKeyUsageRuleList>
</cpix:CPIX>
```

## Package your content

Now you have retrieved the relevant documentation, you can package content with Bento4.

The values from the CPIX documents will need to be copied and pasted to the corresponding section in the examples below.

> For more information, please refer to the Bento4 Documentation which can be found at the following link - [Bento4 Documentation](https://www.bento4.com/documentation/mp4dash/)

Reviewer: Okay gunna stop pointing these out now but you get the idea.

Author: Actioned.

### Packaging Content with Widevine and Playready DRM

> Below is an example of how we can send a request to Bento4 to encrypt our content with Widevine and Playready DRM and package it:

Reviewer: ```suggestion
Below is an example of how we can send a request to Bento4 to encrypt our content with Widevine and PlayReady DRM and package it:
```

Author: Actioned.
